### ISSUE TITLE
Inconsistent behaviour when merging maps

### ISSUE THREAD

**Issue by idrarig** [imdario/mergo#26]
I'm pasting a test case that shows inconsistent behaviour when merging maps. I hope it's clear enough. This problem appeared with commit 65847eaa22876897dd35728eb47c4009972e1b17  that fixed issue #23.

``` go
type typeA struct {
    A *typeB
}

type typeB struct {
    B map[string]typeC
}

type typeC struct{ C string }

func TestInconsistentMergeBehaviourOnMaps(t *testing.T) {
    dst := typeA{
        A: &typeB{
            B: map[string]typeC{"C": typeC{"value"}},
        },
    }
    src := typeA{
        A: &typeB{
            B: map[string]typeC{},
        },
    }
    MergeWithOverwrite(&dst.A.B, src.A.B)
    if dst.A.B["C"].C != "value" {
        t.Error("Bad merge (1)") // No problem here.
    }
    MergeWithOverwrite(&dst, src)
    if dst.A.B["C"].C != "value" {
        t.Error("Bad merge (2)") // Problem here.
    }
}
```

**Comment by imdario** [comment]
I'll check this during next week, as I'm out of office until Monday.

**Comment by imdario** [comment]
First of all, sorry again. Second, I will take care of this this next week.

**Comment by imdario** [comment]
I guess it doesn't matter anymore but I checked this today, almost two years later. Shame on me.

This behavior is confusing but expected. I wrote a test to show what happens with structs, as we try to keep merging behaviors coherent between types.

```go
type typeD struct {
        A *typeE
}

type typeE struct {
        B typeC
}

func TestMergeBehaviourOnStruct(t *testing.T) {
        dst := typeD{
                A: &typeE{
                        B: typeC{"value"},
                },
        }
        src := typeD{
                A: &typeE{
                        B: typeC{},
                },
        }
        Merge(&dst.A.B, src.A.B, WithOverride)
        if dst.A.B.C != "value" {
                t.Error("Bad merge (1)") // No problem here.
        }
        Merge(&dst, src, WithOverride)
        if dst.A.B.C != "value" {
                t.Error("Bad merge (2)") // Problem here.
        }
}
```

This fails in the second mergo too, as dst.A.B.C is "". The main point is the merging scope (undocumented concept, I'm just creating it now). When you merge a map or a struct, we compare each key/field value. If it's non-zero, we override the destination key/field if it's zero. We override it anyway if we use MergeWithOverwrite or Merge with param WithOverride.

When you merge a struct which contains a map or a struct, the comparison is made at field level, the one that holds the map or the struct. If the stated condition is true, we overwrite it as a whole.

If it worked in a previous version I wouldn't assume it was ok. Mergo is very tricky and two years ago was far from complete. Also, the pointed commit isn't related to map or struct merging.

So, as it would make even more trickier to handle this in a more fine-grained way and it works in a consistent in both maps and struct, I'm closing this issue.
